src/osd/ReplicatedBackend.cc

```diff
@@ -1295,6 +1295,14 @@ void ReplicatedBackend::calc_head_subsets(
   uint64_t size = obc->obs.oi.size;
   if (size)
     data_subset.insert(0, size);
+  if (parent->min_peer_features() & CEPH_OSD_PARTIAL_RECOVERY) {
+    const auto it = missing.get_items().find(head);
+    assert(it != missing.get_items().end());
+    data_subset.intersection_of(it->second.clean_regions.get_dirty_regions());
+    dout(10) << "calc_head_subsets " << head
+             << "  data_subset " << data_subset << dendl;
+    return;
```

Comment:
That's not really how this works.  Between version v (on the recovery target) and version v' (the current version), we might have created a clone at version c.  Some of the dirty_regions between v and v' might already exist on that clone on the recovery target (because we already recovered it).  Thus, we'd use clone_subsets for those.  It may not be *necessary*, but you're creating a special case here, and that seems worse to me.  Does you code not work properly if clone_subsets is non-empty?